### ISSUE TITLE
Cannot get Update XML file attribute to work

### ISSUE THREAD

**Issue by abw1023** [rfennell/vNextBuild#279]
### VSTS Extension you are using
FileUtilities/XmlFileUpdateTask

### Where are you running it?
on-premise TFS 2017.2

### Version of Extension/Task
3.1.2

### Expected behaviour and actual behaviour
Trying to set appSetting value in web.config.
<add key="BuildVersion" value=""/> exists in the file.
Stops with [error]Unhandled: Cannot read property 'attributes' of null.  Build log is below with parameters.

2018-03-08T23:43:11.1454938Z ==============================================================================
2018-03-08T23:43:11.1454938Z Task         : Update XML file attribute
2018-03-08T23:43:11.1454938Z Description  : Update an attribute in an Xml file based on XPath filter
2018-03-08T23:43:11.1454938Z Version      : 3.1.2
2018-03-08T23:43:11.1454938Z Author       : Black Marble
2018-03-08T23:43:11.1454938Z Help         : Version: 3.1.2. [More Information](https://github.com/rfennell/vNextBuild/wiki/File-Copier-Tasks/)
2018-03-08T23:43:11.1454938Z ==============================================================================
2018-03-08T23:43:11.4423685Z Param: filename - E:\Agent\_work\2\a/Dev/for-deploy/website/Web.config
2018-03-08T23:43:11.4423685Z Param: recurse - true
2018-03-08T23:43:11.4423685Z Param: xpath - /configuration/appSettings/add[@key='BuildVersion']
2018-03-08T23:43:11.4423685Z Param: attribute - value
2018-03-08T23:43:11.4423685Z Param: value - 20180308.7
2018-03-08T23:43:11.5048781Z ##[error]Unhandled: Cannot read property 'attributes' of null

**Comment by rfennell** [comment]
I think we need a bit more debug information to move this one forward.

Please set the build variable of **system.debug = true** this will give a far more detailed log showing all the input parameters and steps taken. 

Also is can I confirm the BuildVersion key is present in the web.config present prior to running the script? 
i.e. are you trying to update the entry not add it.

**Comment by abw1023** [comment]
Yes, 'add key="BuildVersion" value=""' exists in the web.config.  I tried to put that in my original message but the angle brackets caused it to disappear from the message.  Below is the debug output.

The debug output told me what the problem is:
I have the "recurse" value unchecked, and the first debug line mentioning it confirms that value, but later the value has changed to true. This is an MVC project so there are 2 web.config files, one at the root and one in the Views subdirectory (both are required). Only the root contains the appSetting and should be modified.  For testing purposes I removed the Views\web.config file and then the Update XML worked!

So it is the recurse flag ending up true that is causing the problem.

2018-03-09T15:51:50.3228157Z ##[debug]Evaluating condition for step: 'Set Build Version'
2018-03-09T15:51:50.3228157Z ##[debug]Evaluating: succeeded()
2018-03-09T15:51:50.3228157Z ##[debug]Evaluating succeeded:
2018-03-09T15:51:50.3228157Z ##[debug]=> (Boolean) True
2018-03-09T15:51:50.3228157Z ##[debug]Expanded: True
2018-03-09T15:51:50.3228157Z ##[debug]Result: True
2018-03-09T15:51:50.3228157Z ##[section]Starting: Set Build Version
2018-03-09T15:51:50.3228157Z ==============================================================================
2018-03-09T15:51:50.3228157Z Task         : Update XML file attribute
2018-03-09T15:51:50.3228157Z Description  : Update an attribute in an Xml file based on XPath filter
2018-03-09T15:51:50.3228157Z Version      : 3.1.2
2018-03-09T15:51:50.3228157Z Author       : Black Marble
2018-03-09T15:51:50.3228157Z Help         : Version: 3.1.2. [More Information](https://github.com/rfennell/vNextBuild/wiki/File-Copier-Tasks/)
2018-03-09T15:51:50.3228157Z ==============================================================================
2018-03-09T15:51:50.6040748Z ##[debug]agent.workFolder=E:\Agent\_work
2018-03-09T15:51:50.6040748Z ##[debug]loading inputs and endpoints
2018-03-09T15:51:50.6040748Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_SYSTEMVSSCONNECTION_ACCESSTOKEN
2018-03-09T15:51:50.6040748Z ##[debug]loading ENDPOINT_AUTH_SCHEME_SYSTEMVSSCONNECTION
2018-03-09T15:51:50.6040748Z ##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
2018-03-09T15:51:50.6040748Z ##[debug]loading INPUT_ATTRIBUTE
2018-03-09T15:51:50.6040748Z ##[debug]loading INPUT_FILENAME
2018-03-09T15:51:50.6040748Z ##[debug]loading INPUT_RECURSE
2018-03-09T15:51:50.6040748Z ##[debug]loading INPUT_VALUE
2018-03-09T15:51:50.6040748Z ##[debug]loading INPUT_XPATH
2018-03-09T15:51:50.6040748Z ##[debug]loaded 8
2018-03-09T15:51:50.6040748Z ##[debug]filename=E:\Agent\_work\2\a\Dev\for-deploy\website\Web.config
2018-03-09T15:51:50.6040748Z ##[debug]xpath=/configuration/appSettings/add[@key='BuildVersion']
2018-03-09T15:51:50.6040748Z ##[debug]attribute=value
2018-03-09T15:51:50.6040748Z ##[debug]value=20180309.3
2018-03-09T15:51:50.6040748Z ##[debug]recurse=false
2018-03-09T15:51:50.6040748Z Param: filename - E:\Agent\_work\2\a\Dev\for-deploy\website\Web.config
2018-03-09T15:51:50.6040748Z Param: recurse - true
2018-03-09T15:51:50.6040748Z Param: xpath - /configuration/appSettings/add[@key='BuildVersion']
2018-03-09T15:51:50.6040748Z Param: attribute - value
2018-03-09T15:51:50.6040748Z Param: value - 20180309.3
2018-03-09T15:51:50.6196909Z ##[debug]Looking in folder [E:\Agent\_work\2\a\Dev\for-deploy\website] for files that match pattern [Web.config] using recursion [true]
2018-03-09T15:51:50.6665798Z ##[debug]task result: Failed
2018-03-09T15:51:50.6665798Z ##[error]Unhandled: Cannot read property 'attributes' of null
2018-03-09T15:51:50.6665798Z ##[debug]Processed: ##vso[task.issue type=error;]Unhandled: Cannot read property 'attributes' of null
2018-03-09T15:51:50.6665798Z ##[debug]Processed: ##vso[task.complete result=Failed;]Unhandled: Cannot read property 'attributes' of null
2018-03-09T15:51:50.6821936Z ##[section]Finishing: Set Build Version

**Comment by rfennell** [comment]
Great sounds like you are unblocked then?

**Comment by abw1023** [comment]
No, I need the recurse flag fixed.  I must have both web.config files so I cannot use this task as long as it insists on doing a recursive search.

**Comment by rfennell** [comment]
OK I will look into it

**Comment by rfennell** [comment]
There was a problem with the Boolean parameter parsing, this has been fixed in 3.2.2. I also revised the unit tests

Can you confirm this works for you

**Comment by abw1023** [comment]
I appear to only have permission to install from the Marketplace, so I can't verify this fix until it's published.  I'm sure if the unit tests now check it, it will be fine. Thank you for your quick response.

**Comment by rfennell** [comment]
OK I will close the issue, re-open it if you find the problem is not fixed.

The summary of the problem was that there was meant to be a unit test to check recursion processing, but it checked the ProcessFiles() function, this took the recurse parameter as a boolean, and all was OK when it was run (though I did improve the checking on the test). The issue was that the recurse parameter is passed as a string by VSTS into the tasks and the conversion of this value was failing. 

I have now moved the conversion inside the ProcessFiles() function so it can be tested by the existing unit tests - much more robust
